File: apps/shared/ServerGlassUI/HostStore.swift
import Foundation
import ServerGlassFFI

/// Persistence for the servers a person has added.
///
/// Until this existed, adding a server was pointless: the list lived only in memory, so closing the
/// app threw it away and every launch started from the empty state.
///
/// Two stores, deliberately:
///
/// - **Configuration** — address, port, username, which sign-in method, key path — is not secret.
///   It goes in `UserDefaults` as JSON, where it can be inspected and backed up like any other
///   preference.
/// - **Secrets** — passwords and key passphrases — go in the Keychain, and only ever there.
///
/// Secret storage is the one place the "core owns all logic" rule is deliberately broken. The
/// Keychain and the Android Keystore are operating-system facilities backed by hardware the app
/// cannot reach from Rust; reimplementing them in the core would mean inventing key management
/// instead of using the one the platform already audits. The core stays stateless about secrets and
/// is handed one per connection.
public enum HostStore {
    private static let key = "sg.hosts.v1"

    /// Everything needed to reconnect, minus the secret.
    public struct SavedHost: Codable, Identifiable, Equatable {
        public var id: String
        public var address: String
        public var port: UInt16
        public var user: String
        public var authKind: String
        public var keyPath: String?
        public var hostKeyPolicy: String
        public var refreshMs: UInt64

        public init(
            id: String = UUID().uuidString,
            address: String, port: UInt16, user: String, authKind: String,
            keyPath: String?, hostKeyPolicy: String, refreshMs: UInt64
        ) {
            self.id = id
            self.address = address
            self.port = port
            self.user = user
            self.authKind = authKind
            self.keyPath = keyPath
            self.hostKeyPolicy = hostKeyPolicy
            self.refreshMs = refreshMs
        }
    }

    public static func load() -> [SavedHost] {
        guard let data = UserDefaults.standard.data(forKey: key) else { return [] }
        return (try? JSONDecoder().decode([SavedHost].self, from: data)) ?? []
    }

    public static func save(_ hosts: [SavedHost]) {
        guard let data = try? JSONEncoder().encode(hosts) else { return }
        UserDefaults.standard.set(data, forKey: key)
    }

    /// Build the config the core wants, pulling the secret out of the Keychain at the last moment.
    ///
    /// The secret is fetched per connection rather than held alongside the rest of the host, so it
    /// exists in memory for as short a time as the language allows.
    public static func config(for host: SavedHost) -> TargetConfig {
        TargetConfig(
            host: host.address,
            port: host.port,
            user: host.user,
            authKind: host.authKind,
            keyPath: host.keyPath,
            secret: Keychain.secret(for: host.id),
            hostKeyPolicy: host.hostKeyPolicy,
            refreshMs: host.refreshMs
        )
    }

    public static func forget(_ host: SavedHost) {
        Keychain.removeSecret(for: host.id)
    }
}

/// The system Keychain, used for exactly one thing: the secret belonging to a saved host.
enum Keychain {
    private static let service = "cloud.lazarev.serverglass"

    static func setSecret(_ secret: String?, for id: String) {
        removeSecret(for: id)
        guard let secret, !secret.isEmpty, let data = secret.data(using: .utf8) else { return }

        let query: [String: Any] = [
            kSecClass as String: kSecClassGenericPassword,
            kSecAttrService as String: service,
            kSecAttrAccount as String: id,
            kSecValueData as String: data,
            // Available once the device has been unlocked, and never synced to another device or
            // into a backup: a server password should not travel with an iCloud restore.
            kSecAttrAccessible as String: kSecAttrAccessibleAfterFirstUnlockThisDeviceOnly,
        ]
        SecItemAdd(query as CFDictionary, nil)
    }

    static func secret(for id: String) -> String? {
        let query: [String: Any] = [
            kSecClass as String: kSecClassGenericPassword,
            kSecAttrService as String: service,
            kSecAttrAccount as String: id,
            kSecReturnData as String: true,
            kSecMatchLimit as String: kSecMatchLimitOne,
        ]
        var item: CFTypeRef?
        guard SecItemCopyMatching(query as CFDictionary, &item) == errSecSuccess,
            let data = item as? Data
        else { return nil }
        return String(data: data, encoding: .utf8)
    }

    static func removeSecret(for id: String) {
        let query: [String: Any] = [
            kSecClass as String: kSecClassGenericPassword,
            kSecAttrService as String: service,
            kSecAttrAccount as String: id,
        ]
        SecItemDelete(query as CFDictionary)
    }
}

File: apps/windows/ServerGlass.Core/HostStore.cs
using System.Text.Json;

namespace ServerGlass.Core;

/// <summary>Everything needed to reconnect to a host, minus the secret.</summary>
public sealed record SavedHost
{
    public string Id { get; init; } = Guid.NewGuid().ToString();
    public string Address { get; init; } = "";
    public ushort Port { get; init; } = 22;
    public string User { get; init; } = "";

    /// <summary><c>agent</c>, <c>key</c>, <c>key_text</c> or <c>password</c>.</summary>
    public string AuthKind { get; init; } = "agent";

    public string? KeyPath { get; init; }
    public string HostKeyPolicy { get; init; } = "strict";
    public ulong RefreshMs { get; init; } = 1000;
}

/// <summary>
/// Persistence for the servers a person has added.
/// </summary>
/// <remarks>
/// <para>
/// Two stores, deliberately, exactly as on the Apple and Android sides:
/// </para>
/// <list type="bullet">
/// <item><b>Configuration</b> — address, port, username, which sign-in method, key path — is not
/// secret. It goes in <c>hosts.json</c>, where it can be inspected and backed up like any other
/// preference.</item>
/// <item><b>Secrets</b> — passwords, key passphrases and pasted key bodies — go to DPAPI through
/// <see cref="ISecretStore"/>, and only ever there. <see cref="SavedHost"/> has no field to put one
/// in, which is the point: it cannot be written to disk in the clear by accident.</item>
/// </list>
/// <para>
/// Until a store existed on the other platforms, adding a server was pointless — the list lived
/// only in memory, so closing the app threw it away and every launch started from the empty state.
/// </para>
/// </remarks>
public sealed class HostStore
{
    /// <summary>Overrides where the store keeps its files. Set by tests; honoured everywhere.</summary>
    public const string DirectoryVariable = "SERVERGLASS_DATA_DIR";

    private static readonly JsonSerializerOptions FileJson = new()
    {
        WriteIndented = true,
        PropertyNamingPolicy = JsonNamingPolicy.CamelCase,
        PropertyNameCaseInsensitive = true,
    };

    private readonly ISecretStore _secrets;

    public HostStore(ISecretStore secrets, string? directory = null)
    {
        _secrets = secrets;
        Directory = directory
            ?? Environment.GetEnvironmentVariable(DirectoryVariable)
            ?? Path.Combine(
                Environment.GetFolderPath(Environment.SpecialFolder.LocalApplicationData),
                "ServerGlass");
    }

    /// <summary>Where the host list and the trusted host keys live.</summary>
    public string Directory { get; }

    public string HostsPath => Path.Combine(Directory, "hosts.json");

    /// <summary>
    /// Where trusted host keys are recorded.
    /// </summary>
    /// <remarks>
    /// Passed to the core rather than left to it. Windows has no <c>$HOME</c> — it has
    /// <c>USERPROFILE</c> — and an app that guesses wrong here records nothing, which means "remember
    /// this server's identity" silently does nothing and a substituted key is accepted on every
    /// later connection. The core creates the containing directory and reports a write it could
    /// not make, rather than discarding it.
    /// </remarks>
    public string KnownHostsPath => Path.Combine(Directory, "known_hosts");

    public IReadOnlyList<SavedHost> Load()
    {
        if (!File.Exists(HostsPath))
        {
            return [];
        }

        try
        {
            return JsonSerializer.Deserialize<List<SavedHost>>(File.ReadAllText(HostsPath), FileJson)
                   ?? [];
        }
        catch (Exception error) when (error is JsonException or IOException)
        {
            // A corrupt or unreadable list must not stop the app from starting. It is reported
            // rather than swallowed, because a discarded error is a silent failure.
            LoadFailed?.Invoke(this, error);
            return [];
        }
    }

    /// <summary>Raised when the saved list could not be read. The app shows it; it is not fatal.</summary>
    public event EventHandler<Exception>? LoadFailed;

    public void Save(IReadOnlyList<SavedHost> hosts)
    {
        System.IO.Directory.CreateDirectory(Directory);
        // Written to a temporary file and moved into place, so a crash mid-write cannot leave a
        // half-written list where a complete one used to be.
        var temporary = HostsPath + ".tmp";
        File.WriteAllText(temporary, JsonSerializer.Serialize(hosts, FileJson));
        File.Move(temporary, HostsPath, overwrite: true);
    }

    public string? Secret(string hostId, SecretKind kind = SecretKind.Password) =>
        _secrets.Get(hostId, kind);

    /// <returns><c>false</c> when the credential store refused to keep the secret.</returns>
    public bool SetSecret(string hostId, string? secret, SecretKind kind = SecretKind.Password) =>
        _secrets.Set(hostId, kind, secret);

    /// <summary>Erase everything secret belonging to a host.</summary>
    public void Forget(SavedHost host) => _secrets.Forget(host.Id);

    /// <summary>
    /// Build the config the core wants, pulling the secret out of the credential store at the last
    /// moment.
    /// </summary>
    /// <remarks>
    /// Fetched per connection rather than held alongside the rest of the host, so it exists in
    /// managed memory for as short a time as the language allows.
    /// </remarks>
    public TargetConfig Config(SavedHost host) => new()
    {
        Host = host.Address,
        Port = host.Port,
        User = host.User,
        AuthKind = host.AuthKind,
        KeyPath = host.KeyPath,
        // A pasted key is key material, so it lives beside the passphrase in the credential store
        // rather than in the saved record. Stored under its own name so a key and its passphrase
        // can both exist for the same host.
        KeyText = _secrets.Get(host.Id, SecretKind.KeyText),
        Secret = _secrets.Get(host.Id, SecretKind.Password),
        HostKeyPolicy = host.HostKeyPolicy,
        KnownHostsPath = KnownHostsPath,
        RefreshMs = host.RefreshMs,
    };
}
